services/mesh.c

```diff
@@ -1290,6 +1290,24 @@ mesh_send_reply(struct mesh_state* m, int rcode, struct reply_info* rep,
 				&r->edns, &r->query_reply, m->s.region, &r->start_time))
 					r->edns.opt_list_inplace_cb_out = NULL;
 		}
+		/* Send along EDE BOGUS EDNS0 option when answer is bogus */
+		if(rcode == LDNS_RCODE_SERVFAIL &&
+			m->s.env->need_to_validate && (!(r->qflags&BIT_CD) ||
+			m->s.env->cfg->ignore_cd) && rep &&
+			(rep->security <= sec_status_bogus ||
+			rep->security == sec_status_secure_sentinel_fail)) {
+
+			char *reason = m->s.env->cfg->val_log_level >= 2
+			             ? errinf_to_str_bogus(&m->s) : NULL;
+			sldns_ede_code reason_bogus = rep->reason_bogus != LDNS_EDE_DNSSEC_BOGUS
+			                            ? rep->reason_bogus : errinf_to_reason_bogus(&m->s);
+
+			if (m->s.env->cfg->do_ede) {
+				edns_opt_list_append_ede(&r->edns.opt_list_out, m->s.region,
+					reason_bogus, reason);
+			}
```

Comment:
The d-ede test can go one level up, so you don't have to do the work (allocation of reason_bogus in `errinf_to_readon_bogus()`) when it is discarded anyway.
```suggestion
		if(m->s.env->cfg->do_ede && rcode == LDNS_RCODE_SERVFAIL &&
			m->s.env->need_to_validate && (!(r->qflags&BIT_CD) ||
			m->s.env->cfg->ignore_cd) && rep &&
			(rep->security <= sec_status_bogus ||
			rep->security == sec_status_secure_sentinel_fail)) {

			char *reason = m->s.env->cfg->val_log_level >= 2
			             ? errinf_to_str_bogus(&m->s) : NULL;
			sldns_ede_code reason_bogus = rep->reason_bogus != LDNS_EDE_DNSSEC_BOGUS
			                            ? rep->reason_bogus : errinf_to_reason_bogus(&m->s);

			edns_opt_list_append_ede(&r->edns.opt_list_out, m->s.region,
					reason_bogus, reason);
```